Plugin does not work with alfresco multi tenancy

Hello, I am trying to integrate alfresco in a multi tenant environment with onlyoffice document server. Before creating tenants the integration worked but inside a tenant i get strange errors complaining about Node does not exist when trying to save document from Onlyoffice to Alfresco:
```
 2021-05-17 11:52:35,397  DEBUG [parashift.onlyoffice.Prepare] [http-apr-8080-exec-111] Sending JSON prepare object
 2021-05-17 11:52:35,397  DEBUG [parashift.onlyoffice.Prepare] [http-apr-8080-exec-111] {
   "config": {
      "document": {
         "fileType": "docx",
         "key": "f43dbc30-3558-441b-9f8b-3e7b2ad26e90_1.1",
         "permissions": {"edit": true},
         "title": "Document1.docx",
         "url": "http://192.168.77.42:8080/alfresco/s/api/node/content/workspace/SpacesStore/f43dbc30-3558-441b-9f8b-3e7b2ad26e90?alf_ticket=TICKET_f7982288ac8fc863bad700d6e5a3e2606a3bd7a8"
      },
      "documentType": "text",
      "editorConfig": {
         "callbackUrl": "http://192.168.77.42:8080/alfresco/s/parashift/onlyoffice/callback?nodeRef=workspace://SpacesStore/f43dbc30-3558-441b-9f8b-3e7b2ad26e90&cb_key=UL26rgh3CA6w6GN8W5PtaKaM1CwRRkDJOjMsPlnD-9Y",
         "customization": {"forcesave": true},
         "lang": "en-US",
         "mode": "edit",
         "user": {
            "firstname": "Administrator",
            "id": "admin@tenant.info",
            "lastname": "",
            "name": "Administrator "
         }
      },
      "height": "100%",
      "type": "desktop",
      "width": "100%"
   },
   "mime": "application/vnd.openxmlformats-officedocument.wordprocessingml.document",
   "onlyofficeUrl": "http://192.168.77.43/"
}
 2021-05-17 11:52:36,580  DEBUG [parashift.onlyoffice.CallBack] [http-apr-8080-exec-148] Received JSON Callback
 2021-05-17 11:52:36,580  DEBUG [parashift.onlyoffice.CallBack] [http-apr-8080-exec-148] {
   "actions": [{
      "type": 1,
      "userid": "admin@tenant.info"
   }],
   "key": "f43dbc30-3558-441b-9f8b-3e7b2ad26e90_1.1",
   "status": 1,
   "users": ["admin@tenant.info"]
}
 2021-05-17 11:52:36,590  ERROR [parashift.onlyoffice.CallBack] [http-apr-8080-exec-148] org.alfresco.service.cmr.repository.InvalidNodeRefException: Failed invocation: ReflectiveMethodInvocation: public abstract boolean org.alfresco.service.cmr.repository.NodeService.hasAspect(org.alfresco.service.cmr.repository.NodeRef,org.alfresco.service.namespace.QName) throws org.alfresco.service.cmr.repository.InvalidNodeRefException,org.alfresco.service.cmr.dictionary.InvalidAspectException; target is of class [com.sun.proxy.$Proxy21]
        at org.alfresco.repo.tenant.MultiTNodeServiceInterceptor.invoke(MultiTNodeServiceInterceptor.java:151)
        at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
        at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:204)
        at com.sun.proxy.$Proxy21.hasAspect(Unknown Source)
        at sun.reflect.GeneratedMethodAccessor140.invoke(Unknown Source)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.alfresco.repo.service.StoreRedirectorProxyFactory$RedirectorInvocationHandler.invoke(StoreRedirectorProxyFactory.java:231)
        at com.sun.proxy.$Proxy43.hasAspect(Unknown Source)
        at sun.reflect.GeneratedMethodAccessor140.invoke(Unknown Source)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:317)
        at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:183)
        at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:150)
        at org.alfresco.repo.tagging.TagScopePropertyMethodInterceptor.invoke(TagScopePropertyMethodInterceptor.java:159)
        at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
        at org.alfresco.repo.node.MLPropertyInterceptor.invoke(MLPropertyInterceptor.java:311)
        at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
        at org.alfresco.repo.node.NodeRefPropertyMethodInterceptor.invoke(NodeRefPropertyMethodInterceptor.java:276)
        at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
        at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:204)
        at com.sun.proxy.$Proxy21.hasAspect(Unknown Source)
        at sun.reflect.GeneratedMethodAccessor140.invoke(Unknown Source)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:317)
        at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:198)
        at com.sun.proxy.$Proxy21.hasAspect(Unknown Source)
        at org.alfresco.repo.coci.CheckOutCheckInServiceImpl.getWorkingCopy_aroundBody12(CheckOutCheckInServiceImpl.java:768)
        at org.alfresco.repo.coci.CheckOutCheckInServiceImpl$AjcClosure13.run(CheckOutCheckInServiceImpl.java:1)
        at org.aspectj.runtime.reflect.JoinPointImpl.proceed(JoinPointImpl.java:221)
        at org.alfresco.traitextender.AJExtender.localProceed(AJExtender.java:729)
        at org.alfresco.traitextender.AJProxyTrait.invoke(AJProxyTrait.java:86)
        at com.sun.proxy.$Proxy83.getWorkingCopy(Unknown Source)
        at org.alfresco.repo.virtual.bundle.VirtualCheckOutCheckInServiceExtension.getWorkingCopy(VirtualCheckOutCheckInServiceExtension.java:164)
        at sun.reflect.GeneratedMethodAccessor1530.invoke(Unknown Source)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.alfresco.traitextender.SingletonExtensionFactory$TraiSingletontHandler.invoke(SingletonExtensionFactory.java:74)
        at com.sun.proxy.$Proxy305.getWorkingCopy(Unknown Source)
        at sun.reflect.GeneratedMethodAccessor1529.invoke(Unknown Source)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.alfresco.traitextender.AJExtender.extendAroundAdvice(AJExtender.java:655)
        at org.alfresco.traitextender.RouteExtensions.ajc$inlineAccessMethod$org_alfresco_traitextender_RouteExtensions$org_alfresco_traitextender_AJExtender$extendAroundAdvice(RouteExtensions.java:1)
        at org.alfresco.traitextender.RouteExtensions.intercept(RouteExtensions.java:85)
        at org.alfresco.repo.coci.CheckOutCheckInServiceImpl.getWorkingCopy(CheckOutCheckInServiceImpl.java:765)
        at com.parashift.onlyoffice.CallBack$ProccessRequestCallback.execute(CallBack.java:194)
        at org.alfresco.repo.transaction.RetryingTransactionHelper.doInTransaction(RetryingTransactionHelper.java:464)
        at com.parashift.onlyoffice.CallBack.execute(CallBack.java:158)
        at org.alfresco.repo.web.scripts.RepositoryContainer.transactionedExecute(RepositoryContainer.java:467)
        at org.alfresco.repo.web.scripts.RepositoryContainer.transactionedExecuteAs(RepositoryContainer.java:656)
        at org.alfresco.repo.web.scripts.RepositoryContainer.executeScriptInternal(RepositoryContainer.java:355)
        at org.alfresco.repo.web.scripts.RepositoryContainer.executeScript(RepositoryContainer.java:308)
        at org.springframework.extensions.webscripts.AbstractRuntime.executeScript(AbstractRuntime.java:399)
        at org.springframework.extensions.webscripts.AbstractRuntime.executeScript(AbstractRuntime.java:210)
        at org.springframework.extensions.webscripts.servlet.WebScriptServlet.service(WebScriptServlet.java:132)
        at javax.servlet.http.HttpServlet.service(HttpServlet.java:727)
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:303)
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:208)
        at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:52)
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:241)
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:208)
        at net.bull.javamelody.MonitoringFilter.doFilter(MonitoringFilter.java:237)
        at net.bull.javamelody.MonitoringFilter.doFilter(MonitoringFilter.java:214)
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:241)
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:208)
        at org.alfresco.module.aosmodule.service.ContextRootFilter.doFilter(ContextRootFilter.java:93)
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:241)
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:208)
        at org.alfresco.web.app.servlet.GlobalLocalizationFilter.doFilter(GlobalLocalizationFilter.java:68)
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:241)
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:208)
        at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:220)
        at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:122)
        at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:504)
        at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:170)
        at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:103)
        at org.apache.catalina.valves.AccessLogValve.invoke(AccessLogValve.java:950)
        at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:116)
        at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:421)
        at org.apache.coyote.http11.AbstractHttp11Processor.process(AbstractHttp11Processor.java:1074)
        at org.apache.coyote.AbstractProtocol$AbstractConnectionHandler.process(AbstractProtocol.java:611)
        at org.apache.tomcat.util.net.AprEndpoint$SocketWithOptionsProcessor.run(AprEndpoint.java:2403)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
        at java.lang.Thread.run(Thread.java:748)
```

OK, so i made some tests and we figured out why we got the Empty Callback in onlyoffice logs. It was due to the configuration of onlyoffice ("rejectUnauthorized":  true).  We made a local.json and we specified "rejectUnauthorized": false. Now the checkout is working for the main alfresco.
But the main problem is still present. For alfresco tenants we could not edit documents in ONLYOFFICE. I am pasting more error logs:

[2021-05-18T10:51:09.671] [ERROR] nodeJS - postData error: docId = fb85fbf7-2eee-4ae4-8690-6f48e859a575_1.0;url = http://172.20.9.24:8080/alfresco/s/parashift/onlyoffice/callback?nodeRef=workspace://SpacesStore/fb85fbf7-2eee-4ae4-8690-6f48e859a575&cb_key=G8HdNuZRa0hGveiYSdw_5wMj5n-nW1zdvErXd3I7JL8;data = {"key":"fb85fbf7-2eee-4ae4-8690-6f48e859a575_1.0","status":1,"users":["admin@tenant.local"],"actions":[{"type":1,"userid":"admin@tenant.local"}]}
Error: Error response: statusCode:500; headers:{"content-length":"406","date":"Tue, 18 May 2021 07:51:12 GMT","connection":"close"}; body:
{"error":1, "message":"Failed invocation: ReflectiveMethodInvocation: public abstract boolean org.alfresco.service.cmr.repository.NodeService.hasAspect(org.alfresco.service.cmr.repository.NodeRef,org.alfresco.service.namespace.QName) throws org.alfresco.service.cmr.repository.InvalidNodeRefException,org.alfresco.service.cmr.dictionary.InvalidAspectException; target is of class [com.sun.proxy.$Proxy39]"}
    at Request._callback (/snapshot/server/build/server/Common/sources/utils.js:0:0)
    at Request.init.self.callback (/snapshot/server/build/server/Common/node_modules/request/request.js:185:22)
    at Request.emit (events.js:198:13)
    at Request.<anonymous> (/snapshot/server/build/server/Common/node_modules/request/request.js:1154:10)
    at Request.emit (events.js:198:13)
    at IncomingMessage.<anonymous> (/snapshot/server/build/server/Common/node_modules/request/request.js:1076:12)
    at Object.onceWrapper (events.js:286:20)
    at IncomingMessage.emit (events.js:203:15)
    at endReadableNT (_stream_readable.js:1143:12)
    at process._tickCallback (internal/process/next_tick.js:63:19)
[2021-05-18T10:52:53.874] [ERROR] nodeJS - postData error: docId = fb85fbf7-2eee-4ae4-8690-6f48e859a575_1.0;url = http://172.20.9.24:8080/alfresco/s/parashift/onlyoffice/callback?nodeRef=workspace://SpacesStore/fb85fbf7-2eee-4ae4-8690-6f48e859a575&cb_key=G8HdNuZRa0hGveiYSdw_5wMj5n-nW1zdvErXd3I7JL8;data = {"key":"fb85fbf7-2eee-4ae4-8690-6f48e859a575_1.0","status":4,"actions":[{"type":0,"userid":"admin@tenant.local"}]}
Error: Error response: statusCode:500; headers:{"content-length":"406","date":"Tue, 18 May 2021 07:52:56 GMT","connection":"close"}; body:
{"error":1, "message":"Failed invocation: ReflectiveMethodInvocation: public abstract boolean org.alfresco.service.cmr.repository.NodeService.hasAspect(org.alfresco.service.cmr.repository.NodeRef,org.alfresco.service.namespace.QName) throws org.alfresco.service.cmr.repository.InvalidNodeRefException,org.alfresco.service.cmr.dictionary.InvalidAspectException; target is of class [com.sun.proxy.$Proxy39]"}
    at Request._callback (/snapshot/server/build/server/Common/sources/utils.js:0:0)
    at Request.init.self.callback (/snapshot/server/build/server/Common/node_modules/request/request.js:185:22)
    at Request.emit (events.js:198:13)
    at Request.<anonymous> (/snapshot/server/build/server/Common/node_modules/request/request.js:1154:10)
    at Request.emit (events.js:198:13)
    at IncomingMessage.<anonymous> (/snapshot/server/build/server/Common/node_modules/request/request.js:1076:12)
    at Object.onceWrapper (events.js:286:20)
    at IncomingMessage.emit (events.js:203:15)
    at endReadableNT (_stream_readable.js:1143:12)
    at process._tickCallback (internal/process/next_tick.js:63:19)

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -2,6 +2,7 @@
 
 ## Added
 - connection to the demo server
+- support alfresco multi tenancy [#83](https://github.com/ONLYOFFICE/onlyoffice-alfresco/issues/83)
 
 ##Removed
 - code that supports "legacy transforms"

--- a/repo/src/main/java/com/parashift/onlyoffice/CallBack.java
+++ b/repo/src/main/java/com/parashift/onlyoffice/CallBack.java
@@ -3,6 +3,7 @@ package com.parashift.onlyoffice;
 import org.alfresco.model.ContentModel;
 import org.alfresco.repo.policy.BehaviourFilter;
 import org.alfresco.repo.security.authentication.AuthenticationUtil;
+import org.alfresco.repo.tenant.TenantContextHolder;
 import org.alfresco.repo.transaction.RetryingTransactionHelper.RetryingTransactionCallback;
 import org.alfresco.service.cmr.coci.CheckOutCheckInService;
 import org.alfresco.service.cmr.repository.ContentData;
@@ -120,17 +121,6 @@ public class CallBack extends AbstractWebScript {
                 }
             }
 
-            NodeRef nodeRef = new NodeRef(request.getParameter("nodeRef"));
-            String hash = null;
-            if (cociService.isCheckedOut(nodeRef)) {
-                hash = (String) nodeService.getProperty(cociService.getWorkingCopy(nodeRef), Util.EditingHashAspect);
-            }
-            String queryHash = request.getParameter("cb_key");
-
-            if (hash == null || queryHash == null || !hash.equals(queryHash)) {
-                throw new SecurityException("Security hash verification failed");
-            }
-
             String username = null;
 
             if (callBackJSon.has("users")) {
@@ -149,10 +139,23 @@ public class CallBack extends AbstractWebScript {
 
             if (username != null) {
                 AuthenticationUtil.clearCurrentSecurityContext();
+                TenantContextHolder.setTenantDomain(AuthenticationUtil.getUserTenant(username).getSecond());
                 AuthenticationUtil.setRunAsUser(username);
             } else {
                 throw new SecurityException("No user information");
             }
+
+            NodeRef nodeRef = new NodeRef(request.getParameter("nodeRef"));
+            String hash = null;
+            if (cociService.isCheckedOut(nodeRef)) {
+                hash = (String) nodeService.getProperty(cociService.getWorkingCopy(nodeRef), Util.EditingHashAspect);
+            }
+            String queryHash = request.getParameter("cb_key");
+
+            if (hash == null || queryHash == null || !hash.equals(queryHash)) {
+                throw new SecurityException("Security hash verification failed");
+            }
+
             Boolean reqNew = transactionService.isReadOnly();
             transactionService.getRetryingTransactionHelper()
                 .doInTransaction(new ProccessRequestCallback(callBackJSon, nodeRef), reqNew, reqNew);
@@ -244,6 +247,7 @@ public class CallBack extends AbstractWebScript {
                     cociService.checkin(wc, null, null, true);
 
                     AuthenticationUtil.clearCurrentSecurityContext();
+                    TenantContextHolder.setTenantDomain(AuthenticationUtil.getUserTenant(lockOwner).getSecond());
                     AuthenticationUtil.setRunAsUser(lockOwner);
 
                     nodeService.setProperty(wc, Util.EditingHashAspect, hash);

